File: querier/internal/manifest/manifest.go
// Package manifest gives the Go CLI read-write access to the global shard
// manifest so registry prefetch can register downloaded shards and cache GC
// can evict rows. The schema matches the Kotlin Manifest exactly; SQLite
// files remain the only contract between the two halves.
package manifest

import (
	"database/sql"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"

	_ "modernc.org/sqlite"
)

type Shard struct {
	ShardID          string
	JarCoordinate    *string
	JarFilename      string
	JarSHA256        string
	ExtractorVersion int
	ShardPath        string
	ShardChecksum    *string
	SizeBytes        int64
	CreatedAt        int64
	LastAccessAt     int64
	Source           string
}

// DigestEntry is one cached jar content hash, valid only while the jar's size
// and mtime are unchanged. It lets repeated indexing/prefetch skip re-hashing
// the (immutable) dependency jars that dominate the warm path.
type DigestEntry struct {
	SizeBytes   int64
	MtimeMillis int64
	SHA256      string
}

type DB struct {
	db *sql.DB
}

// Open opens (creating if needed) the manifest database read-write with the
// same pragmas the Kotlin indexer uses, so concurrent runs stay safe.
func Open(path string) (*DB, error) {
	absolute, err := filepath.Abs(path)
	if err != nil {
		return nil, err
	}
	if err := os.MkdirAll(filepath.Dir(absolute), 0o755); err != nil {
		return nil, err
	}
	slashPath := filepath.ToSlash(absolute)
	if filepath.VolumeName(absolute) != "" && !strings.HasPrefix(slashPath, "/") {
		slashPath = "/" + slashPath
	}
	uri := &url.URL{Scheme: "file", Path: slashPath}
	values := uri.Query()
	values.Add("_pragma", "journal_mode(WAL)")
	values.Add("_pragma", "busy_timeout(5000)")
	values.Add("_pragma", "synchronous(FULL)")
	uri.RawQuery = values.Encode()

	db, err := sql.Open("sqlite", uri.String())
	if err != nil {
		return nil, err
	}
	db.SetMaxOpenConns(1)
	if err := db.Ping(); err != nil {
		db.Close()
		return nil, err
	}
	if err := createSchema(db); err != nil {
		db.Close()
		return nil, err
	}
	return &DB{db: db}, nil
}

func (m *DB) Close() error { return m.db.Close() }

// Find returns the shard row for an id, or nil when absent.
func (m *DB) Find(shardID string) (*Shard, error) {
	rows, err := m.db.Query(selectColumns+" FROM shards WHERE shard_id = ?", shardID)
	if err != nil {
		return nil, fmt.Errorf("query manifest: %w", err)
	}
	defer rows.Close()
	if !rows.Next() {
		return nil, rows.Err()
	}
	shard, err := scanShard(rows)
	if err != nil {
		return nil, err
	}
	return &shard, nil
}

// List returns every shard row ordered by last access (oldest first), the
// order LRU eviction consumes.
func (m *DB) List() ([]Shard, error) {
	rows, err := m.db.Query(
		selectColumns + " FROM shards ORDER BY last_access_at ASC, shard_id ASC",
	)
	if err != nil {
		return nil, fmt.Errorf("query manifest: %w", err)
	}
	defer rows.Close()
	var shards []Shard
	for rows.Next() {
		shard, err := scanShard(rows)
		if err != nil {
			return nil, err
		}
		shards = append(shards, shard)
	}
	return shards, rows.Err()
}

// Register inserts a shard row; an existing row for the same id is left
// untouched (mirrors the Kotlin INSERT OR IGNORE).
func (m *DB) Register(shard Shard) error {
	now := time.Now().Unix()
	_, err := m.db.Exec(
		`INSERT OR IGNORE INTO shards(
		   shard_id, jar_coordinate, jar_filename, jar_sha256, extractor_version,
		   shard_path, shard_checksum, size_bytes, created_at, last_access_at, source
		 ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		shard.ShardID,
		shard.JarCoordinate,
		shard.JarFilename,
		shard.JarSHA256,
		shard.ExtractorVersion,
		shard.ShardPath,
		shard.ShardChecksum,
		shard.SizeBytes,
		now,
		now,
		shard.Source,
	)
	if err != nil {
		return fmt.Errorf("register shard: %w", err)
	}
	return nil
}

// Remove deletes a shard row.
func (m *DB) Remove(shardID string) error {
	if _, err := m.db.Exec("DELETE FROM shards WHERE shard_id = ?", shardID); err != nil {
		return fmt.Errorf("remove shard: %w", err)
	}
	return nil
}

// Touch refreshes last_access_at for the given shard ids.
func (m *DB) Touch(shardIDs []string) error {
	if len(shardIDs) == 0 {
		return nil
	}
	now := time.Now().Unix()
	statement, err := m.db.Prepare("UPDATE shards SET last_access_at = ? WHERE shard_id = ?")
	if err != nil {
		return err
	}
	defer statement.Close()
	for _, shardID := range shardIDs {
		if _, err := statement.Exec(now, shardID); err != nil {
			return fmt.Errorf("touch shard %s: %w", shardID, err)
		}
	}
	return nil
}

// SetLastAccess pins last_access_at to an explicit epoch (used by tests and
// tooling that needs deterministic LRU order).
func (m *DB) SetLastAccess(shardID string, epoch int64) error {
	_, err := m.db.Exec(
		"UPDATE shards SET last_access_at = ? WHERE shard_id = ?", epoch, shardID,
	)
	return err
}

// LoadJarDigests returns the whole jar-digest cache keyed by absolute path.
// Callers validate each entry against the file's current size+mtime before
// trusting the hash; a stale or absent entry simply means "re-hash".
func (m *DB) LoadJarDigests() (map[string]DigestEntry, error) {
	rows, err := m.db.Query(
		"SELECT abs_path, size_bytes, mtime_millis, sha256 FROM jar_digest_cache",
	)
	if err != nil {
		return nil, fmt.Errorf("query jar digests: %w", err)
	}
	defer rows.Close()
	result := map[string]DigestEntry{}
	for rows.Next() {
		var path string
		var entry DigestEntry
		if err := rows.Scan(
			&path, &entry.SizeBytes, &entry.MtimeMillis, &entry.SHA256,
		); err != nil {
			return nil, fmt.Errorf("scan jar digest: %w", err)
		}
		result[path] = entry
	}
	return result, rows.Err()
}

// PutJarDigests upserts cache rows in one transaction. Keys are absolute paths.
func (m *DB) PutJarDigests(entries map[string]DigestEntry) error {
	if len(entries) == 0 {
		return nil
	}
	transaction, err := m.db.Begin()
	if err != nil {
		return err
	}
	statement, err := transaction.Prepare(
		`INSERT OR REPLACE INTO jar_digest_cache(abs_path, size_bytes, mtime_millis, sha256)
		 VALUES (?, ?, ?, ?)`,
	)
	if err != nil {
		transaction.Rollback()
		return err
	}
	defer statement.Close()
	for path, entry := range entries {
		if _, err := statement.Exec(
			path, entry.SizeBytes, entry.MtimeMillis, entry.SHA256,
		); err != nil {
			transaction.Rollback()
			return fmt.Errorf("put jar digest %s: %w", path, err)
		}
	}
	return transaction.Commit()
}

const selectColumns = `SELECT shard_id, jar_coordinate, jar_filename, jar_sha256,
       extractor_version, shard_path, shard_checksum, size_bytes,
       created_at, last_access_at, source`

func scanShard(rows *sql.Rows) (Shard, error) {
	var shard Shard
	var coordinate, checksum, source sql.NullString
	var size, created, accessed sql.NullInt64
	if err := rows.Scan(
		&shard.ShardID, &coordinate, &shard.JarFilename, &shard.JarSHA256,
		&shard.ExtractorVersion, &shard.ShardPath, &checksum, &size,
		&created, &accessed, &source,
	); err != nil {
		return Shard{}, fmt.Errorf("scan manifest row: %w", err)
	}
	if coordinate.Valid {
		shard.JarCoordinate = &coordinate.String
	}
	if checksum.Valid {
		shard.ShardChecksum = &checksum.String
	}
	shard.SizeBytes = size.Int64
	shard.CreatedAt = created.Int64
	shard.LastAccessAt = accessed.Int64
	shard.Source = source.String
	return shard, nil
}

// createSchema mirrors the Kotlin Manifest DDL verbatim.
func createSchema(db *sql.DB) error {
	statements := []string{
		`CREATE TABLE IF NOT EXISTS shards (
		  shard_id TEXT PRIMARY KEY,
		  jar_coordinate TEXT,
		  jar_filename TEXT NOT NULL,
		  jar_sha256 TEXT NOT NULL,
		  extractor_version INTEGER NOT NULL,
		  shard_path TEXT NOT NULL,
		  shard_checksum TEXT,
		  size_bytes INTEGER,
		  created_at INTEGER,
		  last_access_at INTEGER,
		  source TEXT
		)`,
		"CREATE INDEX IF NOT EXISTS idx_shards_sha ON shards(jar_sha256)",
		"CREATE INDEX IF NOT EXISTS idx_shards_coord ON shards(jar_coordinate)",
		// Additive jar-digest cache; keyed by absolute path, validated against
		// size+mtime so unchanged jars skip re-hashing on warm runs.
		`CREATE TABLE IF NOT EXISTS jar_digest_cache (
		  abs_path TEXT PRIMARY KEY,
		  size_bytes INTEGER NOT NULL,
		  mtime_millis INTEGER NOT NULL,
		  sha256 TEXT NOT NULL
		)`,
	}
	for _, statement := range statements {
		if _, err := db.Exec(statement); err != nil {
			return fmt.Errorf("create manifest schema: %w", err)
		}
	}
	return nil
}

File: core/src/main/kotlin/dev/research4jar/indexer/store/Manifest.kt
package dev.research4jar.indexer.store

import dev.research4jar.indexer.Research4JarVersions
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.sql.Connection
import java.sql.DriverManager
import java.time.Instant

data class ManifestShard(
    val shardId: String,
    val jarCoordinate: String?,
    val jarFilename: String,
    val jarSha256: String,
    val shardPath: Path,
    val shardChecksum: String?,
    val sizeBytes: Long?,
)

/**
 * One cached jar content hash, valid only while the jar's size and mtime are
 * unchanged. Lets warm re-indexing skip re-hashing immutable dependency jars.
 */
data class CachedDigest(
    val sizeBytes: Long,
    val mtimeMillis: Long,
    val sha256: String,
)

/**
 * A complete manifest row, mirroring the Go manifest.Shard struct field for
 * field (NULL integers read as 0, NULL source as ""). Added for the
 * registry/cache lifecycle port, whose export and GC passes need every
 * column; [ManifestShard] stays the narrower shape the indexer reuses.
 */
data class ManifestRow(
    val shardId: String,
    val jarCoordinate: String?,
    val jarFilename: String,
    val jarSha256: String,
    val extractorVersion: Int,
    val shardPath: String,
    val shardChecksum: String?,
    val sizeBytes: Long,
    val createdAt: Long,
    val lastAccessAt: Long,
    val source: String,
)

class Manifest(
    private val path: Path,
) {
    init {
        Files.createDirectories(path.parent)
        connect().use(::createSchema)
    }

    fun find(shardId: String): ManifestShard? = connect().use { connection ->
        connection.prepareStatement(
            """
            SELECT shard_id, jar_coordinate, jar_filename, jar_sha256,
                   shard_path, shard_checksum, size_bytes
            FROM shards
            WHERE shard_id = ?
            """.trimIndent(),
        ).use { statement ->
            statement.setString(1, shardId)
            statement.executeQuery().use { result ->
                if (!result.next()) return@use null
                ManifestShard(
                    shardId = result.getString("shard_id"),
                    jarCoordinate = result.getString("jar_coordinate"),
                    jarFilename = result.getString("jar_filename"),
                    jarSha256 = result.getString("jar_sha256"),
                    shardPath = Paths.get(result.getString("shard_path")),
                    shardChecksum = result.getString("shard_checksum"),
                    sizeBytes = result.getLong("size_bytes").let {
                        if (result.wasNull()) null else it
                    },
                )
            }
        }
    }

    /**
     * Returns every shard row ordered by last access (oldest first), the
     * order LRU eviction consumes. Added for the registry/cache lifecycle
     * port; mirrors Go manifest.DB.List.
     */
    fun list(): List<ManifestRow> = connect().use { connection ->
        connection.prepareStatement(
            """
            SELECT shard_id, jar_coordinate, jar_filename, jar_sha256,
                   extractor_version, shard_path, shard_checksum, size_bytes,
                   created_at, last_access_at, source
            FROM shards
            ORDER BY last_access_at ASC, shard_id ASC
            """.trimIndent(),
        ).use { statement ->
            statement.executeQuery().use { result ->
                val rows = mutableListOf<ManifestRow>()
                while (result.next()) {
                    rows += ManifestRow(
                        shardId = result.getString("shard_id"),
                        jarCoordinate = result.getString("jar_coordinate"),
                        jarFilename = result.getString("jar_filename"),
                        jarSha256 = result.getString("jar_sha256"),
                        extractorVersion = result.getInt("extractor_version"),
                        shardPath = result.getString("shard_path"),
                        shardChecksum = result.getString("shard_checksum"),
                        sizeBytes = result.getLong("size_bytes"),
                        createdAt = result.getLong("created_at"),
                        lastAccessAt = result.getLong("last_access_at"),
                        source = result.getString("source") ?: "",
                    )
                }
                rows
            }
        }
    }

    /**
     * Refreshes last_access_at for shards reused from the cache so LRU
     * garbage collection evicts genuinely unused shards first.
     */
    fun touch(shardIds: Collection<String>) {
        if (shardIds.isEmpty()) return
        val now = Instant.now().epochSecond
        connect().use { connection ->
            connection.prepareStatement(
                "UPDATE shards SET last_access_at = ? WHERE shard_id = ?",
            ).use { statement ->
                shardIds.forEach { shardId ->
                    statement.setLong(1, now)
                    statement.setString(2, shardId)
                    statement.addBatch()
                }
                statement.executeBatch()
            }
        }
    }

    fun remove(shardId: String) {
        connect().use { connection ->
            connection.prepareStatement("DELETE FROM shards WHERE shard_id = ?").use { statement ->
                statement.setString(1, shardId)
                statement.executeUpdate()
            }
        }
    }

    /**
     * Inserts a shard row; an existing row for the same id is left untouched.
     * [source] and [extractorVersion] default to the indexer's local-build
     * values; the registry prefetch port passes source="remote" and the
     * registry's extractor version (mirrors Go manifest.DB.Register).
     */
    fun register(
        shardId: String,
        coordinate: String?,
        jarFilename: String,
        jarSha256: String,
        shardPath: Path,
        shardChecksum: String,
        sizeBytes: Long,
        source: String = "local",
        extractorVersion: Int = Research4JarVersions.EXTRACTOR,
    ) {
        val now = Instant.now().epochSecond
        connect().use { connection ->
            connection.prepareStatement(
                """
                INSERT OR IGNORE INTO shards(
                  shard_id, jar_coordinate, jar_filename, jar_sha256, extractor_version,
                  shard_path, shard_checksum, size_bytes, created_at, last_access_at, source
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """.trimIndent(),
            ).use { statement ->
                statement.setString(1, shardId)
                statement.setString(2, coordinate)
                statement.setString(3, jarFilename)
                statement.setString(4, jarSha256)
                statement.setInt(5, extractorVersion)
                statement.setString(6, shardPath.toAbsolutePath().normalize().toString())
                statement.setString(7, shardChecksum)
                statement.setLong(8, sizeBytes)
                statement.setLong(9, now)
                statement.setLong(10, now)
                statement.setString(11, source)
                statement.executeUpdate()
            }
        }
    }

    /**
     * Loads the whole jar-digest cache keyed by absolute path. Callers must
     * re-validate each entry against the file's current size+mtime before
     * trusting the hash; a stale or missing entry just means "re-hash".
     */
    fun loadJarDigests(): Map<String, CachedDigest> = connect().use { connection ->
        connection.prepareStatement(
            "SELECT abs_path, size_bytes, mtime_millis, sha256 FROM jar_digest_cache",
        ).use { statement ->
            statement.executeQuery().use { result ->
                val digests = HashMap<String, CachedDigest>()
                while (result.next()) {
                    digests[result.getString("abs_path")] = CachedDigest(
                        sizeBytes = result.getLong("size_bytes"),
                        mtimeMillis = result.getLong("mtime_millis"),
                        sha256 = result.getString("sha256"),
                    )
                }
                digests
            }
        }
    }

    /** Upserts jar-digest cache rows in one transaction. Keys are absolute paths. */
    fun putJarDigests(entries: Map<String, CachedDigest>) {
        if (entries.isEmpty()) return
        connect().use { connection ->
            val previousAutoCommit = connection.autoCommit
            connection.autoCommit = false
            try {
                connection.prepareStatement(
                    """
                    INSERT OR REPLACE INTO jar_digest_cache(
                      abs_path, size_bytes, mtime_millis, sha256
                    ) VALUES (?, ?, ?, ?)
                    """.trimIndent(),
                ).use { statement ->
                    entries.forEach { (path, digest) ->
                        statement.setString(1, path)
                        statement.setLong(2, digest.sizeBytes)
                        statement.setLong(3, digest.mtimeMillis)
                        statement.setString(4, digest.sha256)
                        statement.addBatch()
                    }
                    statement.executeBatch()
                }
                connection.commit()
            } catch (exception: Exception) {
                connection.rollback()
                throw exception
            } finally {
                connection.autoCommit = previousAutoCommit
            }
        }
    }

    private fun connect(): Connection =
        DriverManager.getConnection("jdbc:sqlite:${path.toAbsolutePath()}").also { connection ->
            connection.createStatement().use { statement ->
                statement.execute("PRAGMA journal_mode=WAL")
                statement.execute("PRAGMA busy_timeout=5000")
                statement.execute("PRAGMA synchronous=FULL")
            }
        }

    private fun createSchema(connection: Connection) {
        connection.createStatement().use { statement ->
            statement.execute(
                """
                CREATE TABLE IF NOT EXISTS shards (
                  shard_id TEXT PRIMARY KEY,
                  jar_coordinate TEXT,
                  jar_filename TEXT NOT NULL,
                  jar_sha256 TEXT NOT NULL,
                  extractor_version INTEGER NOT NULL,
                  shard_path TEXT NOT NULL,
                  shard_checksum TEXT,
                  size_bytes INTEGER,
                  created_at INTEGER,
                  last_access_at INTEGER,
                  source TEXT
                )
                """.trimIndent(),
            )
            statement.execute("CREATE INDEX IF NOT EXISTS idx_shards_sha ON shards(jar_sha256)")
            statement.execute(
                "CREATE INDEX IF NOT EXISTS idx_shards_coord ON shards(jar_coordinate)",
            )
            statement.execute(
                """
                CREATE TABLE IF NOT EXISTS jar_digest_cache (
                  abs_path TEXT PRIMARY KEY,
                  size_bytes INTEGER NOT NULL,
                  mtime_millis INTEGER NOT NULL,
                  sha256 TEXT NOT NULL
                )
                """.trimIndent(),
            )
        }
    }
}
